src/heap.rs

```diff
@@ -57,13 +54,13 @@ impl HeapInterface for Heap {
         }
         result
     }
-    fn memset(&mut self, src: &[u8]) {
-        self.0 = src.to_vec();
-    }
 }
 
 #[derive(Debug, Clone)]
-pub struct Heaps(Vec<Heap>);
+pub struct Heaps {
+    heaps: Vec<Heap>,
+    bootloader_heap_rollback_info: Vec<(u32, U256)>,
```

Comment:
Does this `Vec` only ever grows during VM execution (i.e., never emptied)? Maybe, it makes sense to introduce explicit snapshot commit (or implicit one somehow?) so that we do have an ability to drop unnecessary snapshot info. AFAIU, we do have this hook in the glue code (`pop_snapshot_no_rollback`).

More generally, could the snapshot lifecycle be simplified so that (external) snapshots are never embedded into each other? (I *think* we have this property rn with how VM is used in Era, but not 100% sure.) If this is the case (or even if it's not), it could make sense to embed external snapshot management into `VirtualMachine` so that it maintains basic invariants internally, such as changes being committed / rolled back in the LIFO order.

Reply:
I remember that there was some case with two external snapshots but I don't remember where it was or if it was important.

Reply:
Yep, there are embedded snapshots in `MainBatchExecutor` code:

- [`execute_tx()`](https://github.com/matter-labs/zksync-era/blob/dea6969d1b67c54a0985278de68a8d50f1084dc1/core/node/state_keeper/src/batch_executor/main_executor.rs#L171) takes a snapshot unconditionally and never rolls its back or pops it. IIUC, it's expected that [`rollback_last_tx`](https://github.com/matter-labs/zksync-era/blob/dea6969d1b67c54a0985278de68a8d50f1084dc1/core/node/state_keeper/src/batch_executor/main_executor.rs#L204C8-L204C24) is called after a halted transaction, but it looks concerning that a snapshot doesn't seem to be popped otherwise; I'll investigate whether it's the case.
- [`execute_tx_in_vm_with_optional_compression`](https://github.com/matter-labs/zksync-era/blob/dea6969d1b67c54a0985278de68a8d50f1084dc1/core/node/state_keeper/src/batch_executor/main_executor.rs#L257) takes an additional snapshot to handle compression failures. It rolls back a snapshot or pops it inside the method.

There are no VM modifications between taking these 2 snapshots, which makes me think that it could be possible to avoid embedding here. The only other use of `VmInstance::make_snapshot()` looks to be in [the `vm_utils` crate](https://github.com/matter-labs/zksync-era/blob/dea6969d1b67c54a0985278de68a8d50f1084dc1/core/lib/vm_utils/src/lib.rs#L61), and it follows the linear pattern.

Reply:
It's worth doing probably but I'd say making the change is out of scope for this PR.